[#15] Update README (#20)
tezos-commons/homebase-lite: README.md

# Homebase Lite

[![License: MIT](https://img.shields.io/badge/License-MIT-brightgreen.svg)](https://opensource.org/licenses/MIT) [![Build status](https://badge.buildkite.com/680d303398dd927c109b9c0324e7eacf9c7e2e422210ffb746.svg?branch=master)](https://buildkite.com/serokell/homebase-lite)

Homebase Lite is a Tezos smart contract for a decentralized voting system, based
on “snapshots” of the blockchain.

The contract specification is available in [`docs/specification.md`](docs/specification.md)

## Prerequisites

### Build tools

If you want to build this software from source you will need
[Haskell Tool Stack](https://docs.haskellstack.org/en/stable/install_and_upgrade/)

### Operating system

Please note that only Linux is officially supported.

> If you want to use `tzbtc-client` on MacOS, you can try [building it from source](#build-instructions) using `stack`.

Reviewer: ```suggestion
If you want to use `homebase-lite` on MacOS, you can try [building it from source](#build-instructions) using `stack`.
```

### Executables

> In order to use `homebase-lite` executable or run network tests, you will need to obtain `tezos-client` executable.
> `tezos-client` is used for key storing and operation signing.

Reviewer: ```suggestion
In order to use the `homebase-lite` executable or run network tests, you will need to obtain the `tezos-client` executable, which is used for key storing and operation signing.
```
BTW I propose that we use the word `tool` consistently rather than `executable` and `binary`, but I don't have a particularly strong opinion on this, what do you think?

Reviewer: @lierdakil did you miss the question under this suggestion?

Author: Uh... no, I've seen it, I think I've switched to "tool" in a few places. However, it does start to sound somewhat weird at some point if we're using "tool" everywhere. I've revised it once more, switched to "tool" where it made sense, and left "executable" where it's specifically about executables, and not nebulous "tools". "Binary" is only left in "binary releases".

Reviewer: Oh, my bad, I could have sworn there were some more left.

I see none now :ok_hand:

> Various forms of distribution for `tezos-client` are presented in [tezos-packaging repo](https://github.com/serokell/tezos-packaging).

Reviewer: ```suggestion
Various forms of distribution for `tezos-client` are presented in the [tezos-packaging repo](https://github.com/serokell/tezos-packaging).
```

## Obtaining the binary

You can obtain `homebase-lite` binary either by building from source, or downloading an automatically built static executable from [the releases page](https://github.com/tezos-commons/homebase-lite/releases).

### Build instructions

> This project is using [Haskell Tool Stack](https://docs.haskellstack.org/en/stable/install_and_upgrade/), please follow the link to find instructions on obtaining the `stack` executable.

Reviewer: ```suggestion
This project is using the [Haskell Tool Stack](https://docs.haskellstack.org/en/stable/install_and_upgrade/), please follow the link to find instructions on obtaining the `stack` executable.
```

Once you have `stack`, building should be as easy as cloning the repository and running `stack install` in the root of the cloned tree:

```bash
$ git clone https://github.com/tezos-commons/homebase-lite.git
$ cd homebase-lite
$ stack install
```

## Using the binary

Once you've obtained `homebase-lite` executable,
use `homebase-lite --help` to get a list of available commands.

You can also get help for a specific command by running `homebase-lite COMMAND --help`.

### Getting the Michelson code of the compiled contract

Use `homebase-lite print` to get the Michelson code of the contract, e.g.

```bash
$ homebase-lite print

> ```

Reviewer: This `e.g.` and code block sound a bit redundant here :sweat_smile:

> This will save the contract code to `homebase-lite.tz` in the current working directory. Alternatively, add `--output FILEPATH` key to change the output path, or use `-` to print to standard output:

Reviewer: Is "key" commonly used or considered to be a more proper term? :thinking: 

I mostly see these referred to as arguments, I think.

Author: Hmm. Good point. That's likely a double mistranslation. The proper term here, from which "key" seems to be ultimately derived, is "keyword argument" probably. Other possibilities include "CLI option" or "named argument". The reason I didn't just go with "argument" is because that's colloquially used for positional arguments, at least in my experience, and, well, these aren't positional. What do you reckon?

Reviewer: Ah, I see :thinking: 

Well, I think we can take a page out of the [`optparse-applicative` book](https://hackage.haskell.org/package/optparse-applicative-0.17.0.0/docs/Options-Applicative.html#g:3), whose terminology seems reasonable to me and loosely based on the POSIX standards:
> - Flags: simple no-argument options
> - Options: options with an argument
> - Arguments: positional arguments
> - Commands. A command defines a completely independent sub-parser

and simply refer to these as "options".

>  "argument" is because that's colloquially used for positional arguments, at least in my experience

Good point, come to think of it I'd say the same (and it also agrees with the above).

You could print the contract's Michelson code and initial storage and then just use `tezos-client originate` if you wanted to, but `homebase-lite` executable provides a more convenient option: `homebase-lite originate`. You will need to have `tezos-client` binary available on your system.

You'll need to specify at least FA2 contract address and administrator's keyhash. You also might need to specify Tezos node endpoint URL with `--endpoint` (but if you don't, the one from `tezos-client` config will be used):

```bash
$ homebase-lite originate --fa2-address KT1AbgvM5D1wAVZbsB3WTCB54E94p2hn1Rb9 --admin tz1Znvao83anTh654Uu44Kb9atTp5yTK93A8
```

If tezos-client binary is not in PATH, you will need to specify the full path to the executable with `--client-path` option, e.g.

```bash
$ homebase-lite originate --fa2-address KT1AbgvM5D1wAVZbsB3WTCB54E94p2hn1Rb9 --admin tz1Znvao83anTh654Uu44Kb9atTp5yTK93A8 --client-path /path/to/tezos-client
```

Note that the origination is performed as the admin address, so you can not specify the admin address for an account you don't have the private key for. Also, obviously, the admin account needs enough tz for the origination.

> Please also note there is no way to change the fa2 contract address after the contract is originated, so double-check if it's correct.

Reviewer: ```suggestion
Please also note that there is no way to change the fa2 contract address after the contract is originated, so double-check if it's correct.
```

Upon successful origination, the contract's address is printed to stdout and stored in `tezos-client`'s alias store with the alias `homebase-lite`. Note that if the alias already exists it will be overwritten. You can specify a different alias with the `--name` key.

You can also specify FA2 token id and any other initial storage parameters. See `homebase-lite originate --help` for more information.

## Running the tests

> If you want to run the included automated testsuite, you'll need to first [build from source](#build-instructions).

Reviewer: ```suggestion
If you want to run the included automated test suite, you'll need to first [build from source](#build-instructions).
```

There are two kind of tests: emulated tests and network tests.

To run emulated tests only, in the root of the repository run `stack test`. This will build the tests if needed and run them, printing the report to standard output.

To run the network tests, you will need to setup an implicit account with plenty of tz to run tests, as repeated originations can get quite costly (so you probably want to run those on the testnet). You will also need `tezos-client` binary.

To run network tests only, use the following command:

```bash
$ stack test --ta '--cleveland-mode only-network --cleveland-node-endpoint https://testnet.endpoint.url/ --cleveland-data-dir /path/to/temporary/directory --cleveland-moneybag-secret-key unencrypted:edsk...'

> ```

Reviewer: How about adding a `Makefile` target to make calling this easier?

Author: The issue is, apart from `--cleveland-mode only-network`, the user has to specify the rest anyway. Well, maybe except the tempdir, which could be created in the makefile (but honestly that's a bit annoying in itself, due to the idiosyncrasies of Make and the zoo of various `mktemp`-like commands out there, way easier to do this with a shell script). The only thing we'd be sidestepping here is argument escaping.

Reviewer: Yeahhh... that's fair.

Ok, let's not do that, but can you modify it to a multiline command?
I think it would at least look a little bit less scary that way.

Author: Yeah... that's a little bit tricky, too, because of how `stack` handles `--test-arguments`, one spurious space somewhere and stack will complain about unknown options, which would be rather confusing. Okay, I'll try to figure something out with the makefile.

Author: Okay, I did _something_ that looks arguably less scary.
